src/indexable.cr

```diff
@@ -281,6 +281,69 @@ module Indexable(T)
     self
   end
 
+  # Optimized version of `Enumerable#join` that performs better when
+  # all of the elements in this indexable are strings: the total string
+  # bytesize to return can be computed before creating the final string,
+  # which performs better because there's no need to do reallocations.
+  def join(separator = "")
+    return "" if empty?
+
+    {% if T == String %}
+      _join_strings(separator)
+    {% elsif String < T %}
+      if all?(&.is_a?(String))
+        _join_strings(separator)
+      else
+        super(separator)
+      end
+    {% else %}
+      super(separator)
+    {% end %}
+  end
+
+  private def _join_strings(separator)
```

Comment:
I don't like the convention of using `_name` for private names when crystal already has private methods, just `private def join_strings` is fine.

Reply:
The problem with private methods is that they can still be accessed by subclasses or including types. But yeah, I'll change it.